core/src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -115,6 +117,26 @@ public static ShardPath loadShardPath(ESLogger logger, NodeEnvironment env, Shar
         }
     }
 
+    /**
+     * This method tries to delete left-over shards where the index name has been reused but the UUID is different
+     * to allow the new shard to be allocated.
+     */
+    public static void deleteLeftoverShardDirectory(ESLogger logger, NodeEnvironment env, ShardLock lock, @IndexSettings Settings indexSettings) throws IOException {
+        final String indexUUID = indexSettings.get(IndexMetaData.SETTING_INDEX_UUID, IndexMetaData.INDEX_UUID_NA_VALUE);
+        final Path[] paths = env.availableShardPaths(lock.getShardId());
+        for (Path path : paths) {
+            ShardStateMetaData load = ShardStateMetaData.FORMAT.loadLatestState(logger, path);
+            if (load != null) {
+                if ((load.indexUUID.equals(indexUUID) || IndexMetaData.INDEX_UUID_NA_VALUE.equals(load.indexUUID)) == false) {
```

Comment:
I think it would be clearer to have the negation with each check directly (then parenthesis don't matter)?
```
if (load.indexUUID.equals(indexUUID) == false
    && IndexMetaData.INDEX_UUID_NA_VALUE.equals(load.indexUUID) == false)
```